pkg/gadgets/top/ebpftop/piditer/iter.go

```diff
@@ -0,0 +1,191 @@
+}
+
+type PidIterEntry struct {
+	ProgID uint32
+	Pid    uint32
+	Comm   string
+}
+
+var (
+	addrLock        sync.Mutex
+	bpfProgFopsAddr uint64
+	triedGetAddr    bool
+
+	iterEntrySize = int(unsafe.Sizeof(piditerPidIterEntry{}))
+)
+
+// getBpfProgFopsAddr reads the address of bpf_prog_fops from /proc/kallsyms. It
+// only does this on first call and caches the address.
+func getBpfProgFopsAddr() uint64 {
+	addrLock.Lock()
+	defer addrLock.Unlock()
+
+	if triedGetAddr {
+		return bpfProgFopsAddr
+	}
+
+	triedGetAddr = true
+
+	file, err := os.Open("/proc/kallsyms")
+	if err != nil {
+		return 0
+	}
+	defer file.Close()
+
+	scanner := bufio.NewScanner(file)
+	for scanner.Scan() {
+		line := scanner.Text()
+
+		fields := strings.Fields(line)
+		if len(fields) < 3 {
+			continue
+		}
+
+		if fields[2] == "bpf_prog_fops" {
+			addr, err := strconv.ParseUint(fields[0], 16, 64)
+			if err != nil {
+				return 0
+			}
+
+			// cache addr
+			bpfProgFopsAddr = addr
+			return bpfProgFopsAddr
+		}
+	}
+	return bpfProgFopsAddr
+}
+
+func NewTracer() (iter *PidIter, err error) {
+	p := &PidIter{}
+	defer func() {
+		if err != nil {
+			if p.iter != nil {
+				p.iter.Close()
+			}
+			p.objs.Close()
+		}
+	}()
+
+	spec, err := loadPiditer()
+	if err != nil {
+		return nil, fmt.Errorf("failed to load ebpf program: %w", err)
+	}
+
+	consts := map[string]interface{}{
+		"bpf_prog_fops_addr": getBpfProgFopsAddr(),
+	}
+
+	if err = spec.RewriteConstants(consts); err != nil {
+		return nil, fmt.Errorf("error RewriteConstants: %w", err)
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	opts := ebpf.CollectionOptions{}
+
+	if err = spec.LoadAndAssign(&p.objs, &opts); err != nil {
+		return nil, fmt.Errorf("failed to load ebpf program: %w", err)
+	}
+
+	p.iter, err = link.AttachIter(link.IterOptions{
+		Program: p.objs.GadgetEbpftop,
+		Map:     nil,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to attach iter: %w", err)
+	}
+
+	return p, nil
+}
+
+// DumpPids returns an array of PidIterEntry containing information
+// on which pid (and comm) has an open fd to which eBPF Program ID.
+func (p *PidIter) DumpPids() ([]*PidIterEntry, error) {
+	rc, err := p.iter.Open()
+	if err != nil {
+		return nil, fmt.Errorf("failed to open iter: %w", err)
+	}
+	defer rc.Close()
+
+	res := make([]*PidIterEntry, 0)
+
+	buf := make([]byte, 4096/iterEntrySize*iterEntrySize)
+	for {
+		n, err := io.ReadFull(rc, buf)
+		if err != nil {
+			if errors.Is(err, unix.EAGAIN) {
+				continue
+			}
+		}
+		if n == 0 {
+			break
+		}
+		if n%iterEntrySize != 0 {
+			return nil, fmt.Errorf("invalid format: %d", n)
+		}
+		for i := 0; i < n/iterEntrySize; i++ {
+			entry := (*piditerPidIterEntry)(unsafe.Pointer(&buf[i*iterEntrySize]))
+			res = append(res, &PidIterEntry{
+				ProgID: entry.Id,
+				Pid:    entry.Pid,
+				Comm:   C.GoString((*C.char)(unsafe.Pointer(&entry.Comm[0]))),
+			})
+		}
+	}
+
+	return res, nil
+}
+
+func (p *PidIter) Close() error {
+	err := p.iter.Close()
```

Comment:
I haven't analysed the code yet, but I just ran it to see it working, and it is systematically crashing only on one node when deleting the trace. I am testing it on a multi-node cluster:

```
time="2022-07-27T10:58:32Z" level=info msg="Reconcile trace gadget/ebpftop-lxk2w (gadget ebpftop, node worker)"
time="2022-07-27T10:58:32Z" level=info msg="Gadget ebpftop operation \"start\" on gadget/ebpftop-lxk2w"
time="2022-07-27T10:58:32Z" level=info msg="Gadget completed operation. Trace status will be updated accordingly"
time="2022-07-27T10:58:32Z" level=info msg="Updating new status of trace \"gadget/ebpftop-lxk2w\": state=Started operationError=\"\" operationWarning=\"\" output=<0 characters>"
time="2022-07-27T10:58:32Z" level=info msg="Reconcile trace gadget/ebpftop-lxk2w (gadget ebpftop, node worker)"
time="2022-07-27T10:58:32Z" level=info msg="No annotations. Nothing to do."
time="2022-07-27T11:00:38Z" level=info msg="Reconcile trace gadget/ebpftop-lxk2w (gadget ebpftop, node worker)"
time="2022-07-27T11:00:38Z" level=info msg="Deleting gadget/ebpftop-lxk2w"
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x20b0f79]

goroutine 176 [running]:
github.com/kinvolk/inspektor-gadget/pkg/gadgets/top/ebpftop/piditer.(*PidIter).Close(0x0)
  /gadget/pkg/gadgets/top/ebpftop/piditer/iter.go:182 +0x19
github.com/kinvolk/inspektor-gadget/pkg/gadget-collection/gadgets/top/ebpf.(*Trace).run(0xc0002a6de0)
  /gadget/pkg/gadget-collection/gadgets/top/ebpf/gadget.go:210 +0x36b
created by github.com/kinvolk/inspektor-gadget/pkg/gadget-collection/gadgets/top/ebpf.(*Trace).Start
  /gadget/pkg/gadget-collection/gadgets/top/ebpf/gadget.go:178 +0x56c

```

Reply:
Looks like it used the fallback iterator and I missed checking for that. Thanks!

Reply:
Resolved. Thanks!